community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/pipes/ExpandIntoPipe.scala

```diff
@@ -69,6 +65,63 @@ case class ExpandIntoPipe(source: Pipe,
     }
   }
 
+  private def iterateRelationships(row: ExecutionContext, state: QueryState, fromNode: Node, toNode: Node) = {
+    //check degree and iterate from the node with smaller degree
+    val relTypes = types.types( state.query )
+    val fromDegree = getDegree(fromNode, relTypes, dir, state.query)
+    val toDegree = getDegree(toNode, relTypes, dir.reverse, state.query)
+
+    val (start, end, relationships) = if (fromDegree < toDegree)
+      (fromNode, toNode, state.query.getRelationshipsForIds(fromNode, dir, relTypes))
+    else
+      (toNode, fromNode, state.query.getRelationshipsForIds(toNode, dir.reverse(), relTypes))
+
+    if (toNode == null)
+      Iterator.empty
+    else {
+      new PrefetchingIterator[ExecutionContext] {
+        val connectedRelationships = ArrayBuffer.empty[Relationship]
+
+        override def fetchNextOrNull(): ExecutionContext = {
+          while (relationships.hasNext) {
+            val rel: Relationship = relationships.next()
+            val other = rel.getOtherNode(start)
+            if (end == other) {
+              connectedRelationships.append(rel)
+              return row.newWith2(relName, rel, toName, toNode)
+            }
+          }
+          registerConnection(start, end, connectedRelationships)
+          null
+        }
+      }
+    }.asScala
+  }
+
+  private def getDegree(node: Node, relTypes: Option[Seq[Int]], direction: Direction, query: QueryContext) = {
+    relTypes.map {
+      case rels if rels.isEmpty => query.nodeGetDegree(node.getId, direction)
+      case rels => rels.foldLeft(0)( (acc, rel) => acc + query.nodeGetDegree(node.getId, direction, rel))
+    }.getOrElse(query.nodeGetDegree(node.getId, direction))
+  }
+
+  private def hasNoConnection(node1: Node, node2: Node) = {
```

Comment:
could we combine the two lookups? `hasNoConnection`and `fromCache` both do the same lookup twice.